Python/gc_free_threading.c

```diff
@@ -609,6 +640,25 @@ clear_weakrefs(struct collection_state *state)
 {
     PyObject *op;
     WORKSTACK_FOR_EACH(&state->unreachable, op) {
+        if (PyGen_CheckExact(op) ||
+            PyCoro_CheckExact(op) ||
+            PyAsyncGen_CheckExact(op)) {
+            // Ensure any non-refcounted pointers to cyclic trash are converted
+            // to refcounted pointers. This prevents bugs where the generator is
+            // freed after its function object.
+            PyGenObject *gen = (PyGenObject *)op;
+            _PyInterpreterFrame *frame = (_PyInterpreterFrame *)(gen->gi_iframe);
```

Comment:
Would be nice to refactor out the common parts with visiting thread's stacks (something like the `gc_visit_frame` from my comment above)